File: 2685-count-the-number-of-complete-components/2685-count-the-number-of-complete-components.py
from collections import deque
class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        # building the adj metrics 

        adj = [[] for _ in range(n)]

        for u , v in edges :
            adj[u].append(v)
            adj[v].append(u)
            # cause undirected edges
        
        visited = [False]*n
        complete_components = 0

        for i in range(n):
            if not visited[i]:
                # we perform the traversal (DFS)
                v_count = 0
                e_count = 0
                
                queue =deque([i])
                visited[i] = True

                while queue:
                    curr = queue.popleft()
                    v_count += 1
                    e_count += len(adj[curr])
                    for neighbour in adj[curr]:
                        if not visited[neighbour]:
                            queue.append(neighbour)
                            visited[neighbour] = True
                
                # devide by 2 cause we encounter same edge twice
                e_count //= 2

                # main condifiton E = V*(V-1) / 2

                if e_count ==( v_count * (v_count -1) )// 2:
                    complete_components +=1
        
        return complete_components

File: 2685-count-the-number-of-complete-components/2685-count-the-number-of-complete-components.cpp
class Solution {
public:
    int countCompleteComponents(int n, vector<vector<int>>& edges) {
        vector<vector<int>>adj(n);

        for (const auto &edge : edges){
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
        }

        vector<bool>visited(n , false);
        int complete_components = 0;

        for(int i = 0 ; i < n ; i++){
            if(!visited[i]){
                int v_count = 0;
                int e_count = 0;

                queue<int>q;
                q.push(i);
                visited[i]= true;

                while(!q.empty()){
                    int curr = q.front();
                    q.pop();

                    v_count++;
                    e_count += adj[curr].size();

                    for(int neighbour : adj[curr]){
                        if(!visited[neighbour]){
                            visited[neighbour] = true;
                            q.push(neighbour);
                        }
                    }
                }

                e_count /= 2;
                
                // condition 
                if(e_count == (v_count * (v_count -1)) /2){
                    complete_components++;
                }
            }
        }

        return complete_components;
    }
};